walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,135 @@
+    }
+    return meetingTimes;
+  }
+
+  private void populateFreeTimes (ArrayList<TimeRange> freeTimes, int blockLength) {
+    /* Populates FreeTimes with 1 day of blocks of size blockLength
+     * (i.e. blocklength 15 would mean 12:00am, 12:15am, 12:30am, 12:45am, etc)
+     * Note that blockLength must be a factor of 60 minutes or there may be unexpected behaviour.
+     */
+
+    // blocks holds all possible minutes value in the hour (ex: 0, 15, 30 if blockLength=15)
+    ArrayList<Integer> blocks = new ArrayList<Integer>();
+    for (int i = 0; i < 60; i++) {
+      if (i % blockLength == 0){
+        blocks.add(i);
+      }
+    }
+
+    // Save all TimeRanges
+    for (int i = 0; i < 24; i++) {
+      for (int b : blocks){
+        int startTime = TimeRange.getTimeInMinutes(i, b);
+        freeTimes.add(TimeRange.fromStartDuration(startTime, blockLength));
+      }
+    }
+  }
+
+  private void removeBusyTimes(ArrayList<TimeRange> freeTimes, Collection<Event> events, Collection<String> meetingAttendees, int blockLength) {
+    /* Updates freeTimes in place by removing all TimeRanges which overlap with events where
+     * at least one meeting attendee is attending.
+     */
+    
+    // Consider all events
+    for (Event e : events){
+
+      // Check if the event has at least one meetingAttendee (otherwise just skip the event)
+      ArrayList<String> eventAttendees = new ArrayList<String>(e.getAttendees());
+      eventAttendees.retainAll(meetingAttendees);
+      if (eventAttendees.size() != 0) {
+
+        // Remove all busy time blocks (those in the event) from list of free times
+        TimeRange eventTime = e.getWhen();
+        int numBlocks = eventTime.duration() / blockLength;
+        for (int i = 0; i < numBlocks; i++) {
+          freeTimes.remove(TimeRange.fromStartDuration(eventTime.start() + i * blockLength, blockLength));
+        }
+      }
+    }
+  }
+
+  private void updateMeetingTimes(Collection<TimeRange> meetingTimes, ArrayList<TimeRange> freeTimes, long meetingDuration){
+    /* Update meetingTimes in place by adding blocks of uninterrupted freeTimes of minimum length meetingDuration.
+     * This is done by merging continuous blocks of freeTime into one big range.
+     */
+
+    // Merge consecutive times blocks
+    while (freeTimes.size() != 0) {
+      int i = 1;
+
+      // Run until there are not consecutive (i.e. looping stops when there's a "break)
+      while (i < freeTimes.size() && freeTimes.get(i-1).end() == freeTimes.get(i).start()){
+        i++;
+      }
+
+      // Now we have a potential uninterrupted time block to add
+      TimeRange potentialTime = TimeRange.fromStartEnd(freeTimes.get(0).start(), freeTimes.get(i-1).end(), false);
+
+      // Verify that the timeblock is long enough before adding it as a potential meeting time
+      if (potentialTime.duration() >= meetingDuration){
+        meetingTimes.add(potentialTime);
+      }
+
+      // Move onto the next free times
```

Comment:
Mutations are costly and risky, avoid them whenever you can. In this case you should keep track of the start index (e.g. as proposed above) and not change the input list.